### ISSUE TITLE
Question about thin volume

### ISSUE THREAD

**Issue by kevin-chuang** [jthornber/thin-provisioning-tools#53]
Hi Joe,

Any advise would be much appreciated!!
Do you have better/simple way to achieve what I want to do? Thanks again for your help!!

Requirement:
I'm trying to implement a "flatten" function for a thin snapshot volume in the meanwhile a user is issuing IO to this thin snapshot volume. Flatten means making the thin volume having it's own exclusive(not shared) space. **Actually I want to implement to same function as "rbd flatten" in LVM thin volume.**
My implement logic is trying to do something like "dd if=/dev/vg/thin_snap_volume of=/dev/vg/thin_snap_volume"
**The difficulty is my flattening IO(ex. if I use dd) is racing with user's IO, hence there's no way to use "dd if=/dev/vg/thin_snap_volume of=/dev/vg/thin_snap_volume".**

Implement:
0. As my understanding, what schedule_copy do is operation1.fully copy data from shared block to the newly allocated block, operation2.write the IO from bio to the newly allocated block.
**What I want to do is reusing operation1. fully copy... without operation2, in this way I don't need to take care of racing issue.**

1. add a message to dmsetup, message "flatten <dev_id> <start_addr> <length>"
2. add a function named thin_flatten() to dm-thin.c to handle the case of message "flatten"
3. the thin_flatten() do the following:
  a. allocate a dm_target_io
  b. throw dm_target_io to thin_map.
  c. hook copy_complete(), make copy_complete() not to write bio to newly allocated block.
4. write a simple script to get all mapping of /dev/vg/thin_snap_volume and use a for loop to call 
```dmsetup message /dev/vg/pool 0 "flatten snap_id <mapping>"``` 
to flatten every mapping entry.

**Comment by jthornber** [comment]
The trick for this scenario is to use a mirror target.  There are two of
these: either dm-mirror, which I wrote a long time ago for exactly this
scenario (pvmove), or the newer dm-raid.  I'm not sure which of these two
is best to use; LVM still uses dm-mirror for pvmove so that's probably the
simplest.

The sequence is something like this:

i) suspend the device you wish to copy (src).
ii) create a new dm dev, that has an identical table to the src dev.
iii) create a new dm dev for the dest
iv) load a mirror target into the original src dev (that way ongoing io
gets routed through the mirror).  The mirror should have two legs, the
first pointing at the clone of the src (fulling synchronised) and the
second pointing at the dest (and marked unsynchronised).
v) resume everything
vi) wait for the mirror target to report that everything is in sync.

At this point you can tear down the mirror and use which ever copy you
wish, eg, load the dest table into the original src dev.

**Comment by kevin-chuang** [comment]
Thank Joe, your approach is good.

**Comment by kevin-chuang** [comment]
Another question about thin's metadata, I saw lots of code changes done by you. Any advise would be much appreciated!!

I'm using LVM2.2.02.107 with linux-3.10.1(the thin_target version is "1.8.0").
Now I hit a metadata resize error when I'm growing a thin volume as follows:
**"Online metadata resize for ... is not supported."**
I checked the code in LVM2, it seems like LVM only allow online metadata resize when thin_target verions is >= 1.10.0. **I want to remove the checking about 1.10.0**
I checked the code changes between thin_target 1.8.0 and 1.10.0, 
1. There were many code changes about no space handling, hence I think the reason is you were taking good care of metadata extend(resize) failed cases, is it?
2. Continue questions1, if so, that means I can remove the checking "1.10.0" in LVM2 if I can guarantee there are enough free space for metadata allocation, is it?

Thanks for your time!!
